Add benchmark for Docutils

AA-Turner opened the pull request:
This adds a benchmark of Docutils as an application. I thought a reasonable test load was Docutils' own docs (takes ~4.5-5s on my computer).

I haven't submitted a benchmark before---I don't know the best way of storing the input data, so for speed I copied the documentation into git here (the docs are public domain).

A

mdboom commented on pyperformance/data-files/benchmarks/bm_docutils/run_benchmark.py:
```diff
@@ -0,0 +1,60 @@
+        try:
+            dest = out_root / file.relative_to(doc_root).with_suffix(".html")
+            dest.parent.mkdir(parents=True, exist_ok=True)
+            core.publish_file(source_path=str(file),
```
Is it possible to get `docutils` to use a string or `io.BytesIO` as input and output and read the input files into memory in advance so they aren't part of the measured time?  That would remove the effect of I/O (which can vary considerably) and only measure the CPU-bound parts of the benchmark.

mdboom reviewed (changes requested):
This is a good choice for benchmark -- real-world workload over a non-trivial codebase.

In the interest of repository size, would it be possible to remove the images? I don't think docutils does much with them, other than linking to them (though correct me if I'm mistaken), so maybe we only need to include one blank image and adjust all of the links to point to that.

AA-Turner commented:
I blanked every image file, so the files are still there but empty. I also removed every active `.. include::` directive and various other files that are unneeded.

I moved the I/O to be outwith the timing code, I couldn't think of anything better.

A

AA-Turner commented:
I have no clue what is causing CI to fail, when I ran the `bench_docutils` function localy everything worked fine. The logs are also unhelpful (exit code 1 != 0).

A

mdboom commented:
> I have no clue what is causing CI to fail, when I ran the `bench_docutils` function localy everything worked fine. The logs are also unhelpful (exit code 1 != 0).
> 
> A

I think the clue might be in here:

```
Traceback (most recent call last):
Command failed with exit code 1
  File "/home/runner/work/pyperformance/pyperformance/pyperformance/data-files/benchmarks/bm_docutils/run_benchmark.py", line 57, in <module>
    runner.bench_time_func("docutils", bench_docutils, DOC_ROOT)
  File "/home/runner/work/pyperformance/pyperformance/venv/cpython3.10-6b3a2b4e6fa6-compat-c0d88e07feb9/lib/python3.10/site-packages/pyperf/_runner.py", line 462, in bench_time_func
    return self._main(task)
  File "/home/runner/work/pyperformance/pyperformance/venv/cpython3.10-6b3a2b4e6fa6-compat-c0d88e07feb9/lib/python3.10/site-packages/pyperf/_runner.py", line 427, in _main
    bench = self._worker(task)
  File "/home/runner/work/pyperformance/pyperformance/venv/cpython3.10-6b3a2b4e6fa6-compat-c0d88e07feb9/lib/python3.10/site-packages/pyperf/_runner.py", line 401, in _worker
    run = task.create_run()
  File "/home/runner/work/pyperformance/pyperformance/venv/cpython3.10-6b3a2b4e6fa6-compat-c0d88e07feb9/lib/python3.10/site-packages/pyperf/_worker.py", line 284, in create_run
    self.compute()
  File "/home/runner/work/pyperformance/pyperformance/venv/cpython3.10-6b3a2b4e6fa6-compat-c0d88e07feb9/lib/python3.10/site-packages/pyperf/_worker.py", line 348, in compute
    WorkerTask.compute(self)
  File "/home/runner/work/pyperformance/pyperformance/venv/cpython3.10-6b3a2b4e6fa6-compat-c0d88e07feb9/lib/python3.10/site-packages/pyperf/_worker.py", line 273, in compute
    self.compute_warmups_values()
  File "/home/runner/work/pyperformance/pyperformance/venv/cpython3.10-6b3a2b4e6fa6-compat-c0d88e07feb9/lib/python3.10/site-packages/pyperf/_worker.py", line 255, in compute_warmups_values
    self._compute_values(self.values, args.values)
  File "/home/runner/work/pyperformance/pyperformance/venv/cpython3.10-6b3a2b4e6fa6-compat-c0d88e07feb9/lib/python3.10/site-packages/pyperf/_worker.py", line 72, in _compute_values
    raise ValueError("benchmark function returned zero")
ValueError: benchmark function returned zero
```

In some case during the test run, the new benchmark function is returning a time that is zero.

You could try to reproduce this locally by running:

```
python -u -m pyperformance.tests
```

mdboom reviewed (approved):
Thanks for addressing my concerns. I'm approving this pending CI passing.

mdboom commented on pyperformance/data-files/benchmarks/bm_docutils/run_benchmark.py:
```diff
@@ -0,0 +1,57 @@
+else:
+    Trace.show = lambda message, channel: ...  # don't print to console
+
+DOC_ROOT = Path("data/docs").resolve()
```
```suggestion
DOC_ROOT = (Path(__file__).parent / "data" / "docs").resolve()
```

This seems to fix the failure for me.  Without this, if `pyperformance` is run from any directory other than the one containing this file, the relative path is wrong, therefore it doesn't find any files, and the benchmark ends up taking 0 seconds.

AA-Turner replied:
Of course! Thanks for looking into this, I really appreciate it.

A

mdboom reviewed (approved):
LGTM

AA-Turner commented:
@ericsnowcurrently / @gvanrossum you've both committed recently, if you've any time for a review of this PR I'd appreciate it! Thanks

A

kumaraditya303 commented on pyperformance/data-files/benchmarks/bm_docutils/run_benchmark.py:
```diff
@@ -0,0 +1,57 @@
+    elapsed = 0
+    for file in doc_root.rglob("*.txt"):
+        file_contents = file.read_text(encoding="utf-8")
+        t0 = time.perf_counter_ns()
```
```suggestion
        t0 = pyperf.perf_counter()
```

kumaraditya303 commented on pyperformance/data-files/benchmarks/bm_docutils/run_benchmark.py:
```diff
@@ -0,0 +1,57 @@
+                                    "output_encoding": "unicode",
+                                    "report_level": 5,
+                                })
+        elapsed += time.perf_counter_ns() - t0
```
```suggestion
        elapsed += pyperf.perf_counter() - t0
```

kumaraditya303 commented on pyperformance/data-files/benchmarks/bm_docutils/run_benchmark.py:
```diff
@@ -0,0 +1,57 @@
+
+
+def bench_docutils(loops, doc_root):
+    runs_total_ns = 0
```
If you use `pyperf.perf_counter` then you don't have to do this conversions and that is used by all other benchmarks.

mdboom commented:
@AA-Turner: It would be great to have this. Any chance you have time to address @kumaraditya303's concerns?

AA-Turner commented:
@mdboom / @kumaraditya303 sorry for the delay here, please may you re-review?

A
